libsolidity/analysis/TypeChecker.cpp

```diff
@@ -478,6 +478,13 @@ bool TypeChecker::visit(VariableDeclaration const& _variable)
 	Type const* varType = _variable.annotation().type;
 	solAssert(!!varType, "Variable type not provided.");
 
+	if (_variable.referenceLocation() == VariableDeclaration::Location::Transient)
```

Comment:
Why isn't this check enough though?

https://github.com/ethereum/solidity/blob/ce4be6e1e611cf094d9227c6c5dd843c163081a2/libsolidity/codegen/ContractCompiler.cpp#L595

The idea was to let it through analysis, since for contract variables of value types it should be already implemented and only stop it in places where we try to actually use the information to do things like generate code, generate storage layout, etc.

With this, you shouldn't need  to remove any tests and you should get the AST just fine.

Reply:
Because it should be [here](https://github.com/ethereum/solidity/blob/ce4be6e1e611cf094d9227c6c5dd843c163081a2/libsolidity/codegen/ContractCompiler.cpp#L589). I thought that what you suggested was the plan in the first place, but Daniel said otherwise, and suggested this approach. I agree that adding a check in `initializeStateVariables(...)` would be easier, and more in line with what I thought our plan was in the first place; the only thing I'm a bit afraid of is that we've missed other such cases in the backend.

Reply:
IMO the approach with sticking it into analysis is too heavy handed and just unnecessary. I think Daniel is trying to play it safe in case there's more broken in the PR than it appears, but I really don't think that's the case. The main issue with #15001 was that it had literally zero codegen coverage, not even the simplest positive case, so we missed the problem completely.

I'd still put that block in the codegen, just in the right place (TBH, I still don't see why `visit(VariableDeclaration)` does not trigger it, but I didn't really dig into the code and a simple test will show whether it works or not anyway). If you're afraid of missing something, #15217 would be a compromise, though I still don't think we need to go that far. We just need some basic tests for cases we're worried about  :)

Reply:
I can push another commit tomorrow and we can compare, I'm fine with either approach.

Reply:
It's done; @ekpyron @cameel compare the two commits and decide which approach is preferable.

Reply:
The contract compiler visit generates getters for public state variables, which just won't happen for non-public ones. Isn't this sufficient indication that it's unreliable to just assert in codegen until we actually try to implement it (which would involve proper semantics tests that'd make this safe)? Even independently of transient storage the general idea of analysis in general is that nothing that fails analysis should fail codegen (there is some exceptions to that, but those shouldn't be taken as excuse to make this worse)... we did want to split things up into steps and start by parsing from the very beginning and then trigger unimplemented errors - I don't know where the idea was coming from that unimplemented errors should or can only be issued in codegen...

Asserting at initialization of state variables should *probably* cover things - but did you check via-IR codegen? In principle there is loads of cases in which you *refer* to state variables as well, in which I have no idea whether there are any assertions in place currently, but I'd guess there aren't, which is more potential loopholes, even if we can't trigger them (which may only mean that they're not *easy* to trigger rather than it not being possible to trigger them).

Having an honest error in analysis (this stuff *is* as of now unimplemented), is definitely the better choice until we have a codegen PR that actually makes it possible to even tell, if all cases are caught or not.

Reply:
Yup, you're right, `--via-ir` compiles fully and generates bytecode (forgot to check yesterday). First option it is then; I'll still add the assertions in codegen.

> the only thing I'm a bit afraid of is that we've missed other such cases in the backend.

See @cameel, I was right :D